### ISSUE TITLE
Unnecessary loss of precision when computing loss functions

### ISSUE THREAD

**Issue by ariasanovsky** [coreylowman/dfdx#872]
When [computing cross entropy](https://github.com/coreylowman/dfdx/blob/9a208d928b92dd77d4b2fa79144bdc8b872aa7f0/src/losses.rs#L84-L86),

```rust
pub fn cross_entropy_with_logits_loss<S: Shape, E: Dtype, D: Device<E>, T: Tape<E, D>>(
    logits: Tensor<S, E, D, T>,
    target_probs: Tensor<S, E, D>,
) -> Tensor<Rank0, E, D, T> {
    let inv_last_axis_numel = 1.0 / <S as HasAxes<S::LastAxis>>::size(logits.shape()) as f64;
    let probs = logits.log_softmax::<S::LastAxis>();
    (probs * target_probs).mean().negate() / inv_last_axis_numel
}
```

of a tensor along along an axis of length $a$, we calculate $1/a$ as a `f64` and later use it to renormalize the `mean()` values from the hadamard product of `probs` and `target_probs`. This results in [more instructions](https://rust.godbolt.org/z/PdKEod4fq) and an unnecessary loss of precision.

This could be replaced by

```rust
pub fn cross_entropy_with_logits_loss<S: Shape, E: Dtype, D: Device<E>, T: Tape<E, D>>(
    logits: Tensor<S, E, D, T>,
    target_probs: Tensor<S, E, D>,
) -> Tensor<Rank0, E, D, T> {
    let last_axis_numel = <S as HasAxes<S::LastAxis>>::size(logits.shape()) as f64;
    let probs = logits.log_softmax::<S::LastAxis>();
    (probs * target_probs).mean().negate() * last_axis_numel
}
```

This pattern is repeated in `kl_div_with_logits_loss`.

**Comment by ariasanovsky** [comment]
We may also be able to replace the call to `mean()` with a call to `sum()`, but it requires fiddling with the tensor dimensions. Using `mean()` and rescaling by the last axis size calculates a sum of the form $$\text{mean} \cdot \text{last axis numel} = \left(\dfrac{1}{|A||I|}\sum_{(a, i)\in A\times I}p_{a, i}\right)\cdot |A|$$ whereas we could instead compute $$\text{sum} / \text{foo} = \left(\sum_{(a, i)\in A\times I}p_{a, i}\right)\cdot |I|^{-1}.$$

I couldn't immediately see how to compute for $|I|$ using the generic parameter `S` or its axis using the existing trait bounds.
